src/cluster/shard/shard.go

```diff
@@ -209,77 +220,99 @@ func NewShardsFromProto(shards []*placementpb.Shard) (Shards, error) {
 }
 
 type shards struct {
-	shardsMap map[uint32]Shard
+	shards   []Shard
+	shardMap map[uint32]Shard
 }
 
-func (ss shards) All() []Shard {
-	shards := make([]Shard, 0, len(ss.shardsMap))
-	for _, shard := range ss.shardsMap {
-		shards = append(shards, shard)
-	}
-	sort.Sort(SortableShardsByIDAsc(shards))
+func (ss *shards) All() []Shard {
+	shards := make([]Shard, len(ss.shards))
+	copy(shards, ss.shards)
+
 	return shards
 }
 
-func (ss shards) AllIDs() []uint32 {
-	ids := make([]uint32, 0, len(ss.shardsMap))
-	for _, shard := range ss.shardsMap {
-		ids = append(ids, shard.ID())
+func (ss *shards) AllIDs() []uint32 {
+	shardIDs := make([]uint32, 0, len(ss.shards))
+	for _, shrd := range ss.shards {
+		shardIDs = append(shardIDs, shrd.ID())
 	}
-	sort.Sort(SortableIDsAsc(ids))
-	return ids
+
+	return shardIDs
 }
 
-func (ss shards) NumShards() int {
-	return len(ss.shardsMap)
+func (ss *shards) NumShards() int {
+	return len(ss.shards)
 }
 
-func (ss shards) Shard(id uint32) (Shard, bool) {
-	shard, ok := ss.shardsMap[id]
-	return shard, ok
+func (ss *shards) Shard(id uint32) (Shard, bool) {
+	shard, ok := ss.shardMap[id]
+	if !ok {
+		return nil, false
+	}
+
+	return shard, true
 }
 
-func (ss shards) Add(shard Shard) {
-	ss.shardsMap[shard.ID()] = shard
+func (ss *shards) Add(shard Shard) {
+	id := shard.ID()
+	i := sort.Search(len(ss.shards), func(i int) bool { return ss.shards[i].ID() >= id })
+	if i < len(ss.shards) && ss.shards[i].ID() == id {
+		ss.shards[i] = shard
+		ss.shardMap[id] = shard
+		return
+	}
+
+	ss.shards = append(ss.shards, shard)
```

Comment:
Nit: maybe worth a comment explaining this? i.e. "insert into the middle by adding to the end and then shifting data".

Reply:
Given that shards are added relatively rarely, might simply add to the end and sort.

Reply:
I think the small complexity add is worth the fact that won't surprise you with a slowdown of an order of magnitude.